[sw, dif_aes] Introduce a set of AES changes

silvestrst opened the pull request:
The aim of this PR is to merge in the main library with other TODOs (**described in #4833 **) addressed in follow-ups.

Please have a look at the #2382 , for the questions and suggestions that I have gathered whilst working my way through the [DIF AES OpenTitan documentation](https://docs.opentitan.org/hw/ip/aes/doc/).

Key points:

- AES API is relatively straight forward, however there is a number of internal conditions that have to be met for correct operation of the peripheral. These are fundamental checks, and should be done at the DIF level rather than the driver level (in my opinion).

- It does not seem practical to bring the input or output data pointers into alignment as it is done in HMAC and SPI. The data that AES operates on is 128-bits wide, and it would be costly to bring it into the alignment for every 128-bit read/write. I think it should be documented that for maximal efficiency the input/output buffers have to be aligned.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+ * AES configuration data.
+ */
+typedef struct dif_aes_config {
+  dif_aes_key_t key_len;
+  dif_aes_mode_t cipher_mode;
+  dif_aes_operation_t operation;
+  dif_aes_operation_mode_t operation_mode;
```
I was itching to store these values in `dif_aes` after configuration (caching the current configuration for DIF internal use), to avoid a number of mmio operations. I see this more like in dif_spi where rx sizes are stored in `dif_aes`. However, for now I am checking the configuration values internally through mmio  operations. What do you think?

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+ * @return `dif_aes_in_data_set_result_t`.
+ */
+dif_aes_in_data_set_result_t dif_aes_in_data_set(const dif_aes_t *aes,
+                                                 const void *data);
```
Sure, but wouldn't it place additional requirements on the caller - forcing the buffers that they feed to the function to be of `uint32_t` type to prevent aliasing issues?

If it's `void*` and internally we use `memcpy`, then compiler should be able to figure out when the buffer is aligned and do the copy efficiently?

lenary replied:
We decided it was better to reflect that AES is handling 4 32-bit integers, rather than to just have this parameter be a buffer. 

One of the major things we didn't like was that `void*` has an implicit length, something the struct makes much more explicit, which we think is a lot safer.

silvestrst replied:
`char *example_message[19] = "123412341234123412" // 19 bytes with \0`
How would you propose to feed the handle `example_message` into `somefunc`, given:
`typedef struct somestruct { uint32_t someval[4]; } somestruct_t;`
`void somefunc(somestruct_t input_data);`

silvestrst replied:
Done.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+ * @return `dif_aes_out_data_read_result_t`.
+ */
+dif_aes_out_data_read_result_t dif_aes_out_data_read(const dif_aes_t *aes,
+                                                     void *data);
```
Please see my question above.

silvestrst replied:
Done.

silvestrst commented:
> The bulk of the review boils down to replacing the individual behavior setters with combined start functions.

I was thinking about that as well, sounds good. I am planning to do some more AES work next week, so I will address the issues then - and also with that add test cases and implementation.

@lenary and @gkelly thank you for the review.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+/**
+ * AES key length in bits.
+ */
+typedef enum dif_aes_key {
```
Done.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+                          requested operation. */
+} dif_aes_key_set_result_t;
+
+/**
+ * Sets AES Initial Key.
+ *
+ * The peripheral must be in IDLE state for this operation to take effect. These
+ * registers must be written at least once regardless of the AES operation mode
+ * (automatic or manual).
+ * [TODO - Open question Nr. 13 in #2382]
+ *
+ * @param aes AES state data.
+ * @param key AES Initial Key.
+ * @return `dif_aes_key_set_result_t`.
+ */
+dif_aes_key_set_result_t dif_aes_key_set(const dif_aes_t *aes,
+                                         const uint8_t *key);
```
I do think this is useful to keep. For example in `ECB` mode as far as I understand we should not use the same key for more than one block of data. I think the intended use-case is to change the key for every block of input data in ECB mode, then setting up a number of register fields that don't need to be changed is wasteful.

vogelpi replied:
Actually, ECB should not be used at all for the reason mentioned by @silvestrst .

lenary replied:
You would only be using ECB mode if you were encrypting/decrypting a single block, right? Then we may want to slightly modify the ECB's start function, but this function has to be deleted, as it doesn't fit with the transactional nature of the API we requested from the design review.

If we don't want to expose ECB at all through the DIF that's a different issue.

silvestrst replied:
> You would only be using ECB mode if you were encrypting/decrypting a single block, right?

I am not sure that the use-cases are limited to a single block, but maybe they are, I am no expert.

> Then we may want to slightly modify the ECB's start function, but this function has to be deleted, as it doesn't fit with the transactional nature of the API we requested from the design review.

There could be things that you did not consider in the design review, and hence this thread.

> If we don't want to expose ECB at all through the DIF that's a different issue.

It seems that it aligns with what @vogelpi said, however, if it should not be used at all, should we eliminate it from the peripheral completely?

silvestrst replied:
Done.

vogelpi replied:
Sorry to bother you again. I am not fully familiar with the DIFs but I just wanted to confirm that really higher-level software should not use ECB at all, but that we need it at least for DV purposes. If there is a DV for ECB, it should be documented that is should not be used other than for DV.

Having the new ECB-specific function sounds reasonable.

silvestrst replied:
> Sorry to bother you again.

No bother at all, in fact thank you for your help!

silvestrst replied:
> If there is a DV for ECB, it should be documented that is should not be used other than for DV.

I put a note saying that it is discouraged to use this mode, but maybe it needs stronger wording.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+ * @param config AES configuration data.
+ * @return `dif_aes_config_result_t`.
+ */
+dif_aes_config_result_t dif_aes_reconfigure(const dif_aes_t *aes,
+                                            const dif_aes_config_t *config);
```
Done.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+                         irrelevant. */
+} dif_aes_iv_set_result_t;
+
+/**
+ * Sets Initialisation Vector (IV).
+ *
+ * The peripheral must be in IDLE state for this operation to take effect. In
+ * CBC and CTR Cipher modes, these registers must be written at least once when
+ * operating in automatic mode.
+ * [TODO - is it different for the manual mode? Open question Nr. 13 in #2382]
+ * The meaning of the Initialisation Vector (IV) depends on the AES Cipher mode:
+ *  - ECB: always ignored.
+ *  - CBC: Initialisation Vector.
+ *  - CTR: Initial Counter Value.
+ *
+ * @param aes AES state data.
+ * @param iv AES Initialisation Vector/Initial Counter Value.
+ * @return `dif_aes_iv_set_result_t`.
+ */
+dif_aes_iv_set_result_t dif_aes_iv_set(const dif_aes_t *aes, const uint8_t *iv);
```
Done.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+ * @param data AES Input Data.
+ * @return `dif_aes_in_data_set_result_t`.
+ */
+dif_aes_in_data_set_result_t dif_aes_in_data_set(const dif_aes_t *aes,
```
Done.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+ * @param data AES Output Data.
+ * @return `dif_aes_out_data_read_result_t`.
+ */
+dif_aes_out_data_read_result_t dif_aes_out_data_read(const dif_aes_t *aes,
```
Done.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+ * @param aes AES state data.
+ * @return `dif_aes_config_result_t`.
+ */
+dif_aes_config_result_t dif_aes_init(mmio_region_t base_addr,
+                                     const dif_aes_config_t *config,
+                                     dif_aes_t *aes);
```
Done.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+/**
+ * AES configuration data.
+ */
+typedef struct dif_aes_config {
+  dif_aes_key_t key_len;
+  dif_aes_mode_t cipher_mode;
+  dif_aes_operation_t operation;
+  dif_aes_operation_mode_t operation_mode;
+} dif_aes_config_t;
```
I decided to use this structure as a parameter for `dif_aes_<mode>_start` functions "- cipher_mode". It is not used in `init` anymore.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+/**
+ * AES cipher modes.
+ */
+typedef enum dif_aes_mode {
+  kDifAesModeECB = 0, /**< The Electronic Codebook Mode. */
+  kDifAesModeCBC,     /**< The Cipher Block Chaining Mode. */
+  kDifAesModeCTR,     /**< The Counter Mode. */
+} dif_aes_mode_t;
```
Done.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+ * @param trigger AES trigger to control the peripheral in manual mode.
+ * @return `dif_aes_trigger_set_result_t`.
+ */
+dif_aes_trigger_set_result_t dif_aes_trigger_set(const dif_aes_t *aes,
```
Done.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,359 @@
+                             operation. */
+} dif_aes_op_mode_set_result_t;
+
+/**
+ * Change AES operation mode (between automatic and manual). The peripheral must
+ * be in IDLE state for this operation to take effect.
+ *
+ * @param aes AES state data.
+ * @param mode AES operation mode (automatic/manual).
+ * @return `dif_aes_op_mode_set_result_t`.
+ */
+dif_aes_op_mode_set_result_t dif_aes_operation_mode_set(
+    const dif_aes_t *aes, dif_aes_operation_mode_t mode);
```
Done.

silvestrst commented:
I have addressed all of the review comments, and have completely refactored the API as suggested. The implementation is completed as well, however has not been tested yet. Some of the implementation details probably could be further improved, like for example: CBC and CTR start functions can be probably internally split out into a common static function. However, I think having separate public API for these functions is a correct choice.

I think this PR needs at least one more high level header pass, and if there are not too many interface changes - I will promote this PR from "draft" to "open", and add other reviewers.

silvestrst commented:
@vogelpi Thank you for your review, much appreciated. I will wait until @lenary and @gkelly have a chance to run through this PR again, and will fix up all outstanding comments in one go.

lenary commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,377 @@
+} dif_aes_operation_t;
+
+/**
+ * AES operation mode.
```
Yeah, this is a complex one. I'm not sure how we should solve this, but I agree it is confusing. We might just have to use enable/disable and communicate this with the parameter name. I'm not sure `gear` or `transmission` get the right points across - maybe something like `algorithm_advance` or something? (where the options are `auto` or `manual`)

lenary commented:
ECB mode changes can come in a follow-up PR (as should any other mode additions), I feel. Let's see if any of pirmin's comments should also go into a follow-up, or if they're small enough to fix quickly.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,377 @@
+/**
+ * Resets an instance of AES.
+ *
+ * TODO - not certain what exactly this function must do.
```
Done.

silvestrst reviewed (commented):
I have addressed all of the comments apart from (which I intend to resolve in a follow-up):
> The DIF currently provides two separate functions for providing new data (load) and reading the output data. These functions check if new data can be loaded or if new output data is available. This is how these generic functions should work. However, when going for maximum performance, the input ready bit does not need to be checked anymore after the second block as illustrated in the block operation section in the technical spec. This allows you to spare 1 out of 10 peripheral accesses which will probably be significant when processing large messages. I think the DIF should provide either a function for doing large transfers directly, or a separate load function that does not check the input ready bit (with a note that drivers should use this one to implement block operation as mentioned in the spec). Maybe this collides a bit with other requirements/guidelines on DIFs and we might need to find another way. We can also discuss this.

Additionally, I would need to have a look at it closely - as if the speed considerations are so important there might be other boosters apart from an `if` check. At a glance - this `if` alone is not a major performance booster.

silvestrst commented:
Still need to comment the data types, and refactor to match the DIF template that we have.

vogelpi commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,582 @@
+ *
+ * @param aes AES state data.
+ * @param transaction Configuration data.
+ * @paran key_share Masked AES key.
```
@paran -> @param?

vogelpi replied:
it's still @paran. It seems that some changes didn't get committed/pushed yet.

vogelpi reviewed (commented):
Quite some time has passed since starting this PR, and in the meantime quite some changes have been done in the hardware. Some more will land in a couple of days. I have thus done a full review and added pointers to relevant PRs. Sorry for the many comments.

Three main things to consider. We don't need to do this in this PR but at least for follow up PRs:
1. We now have 2 more block cipher modes of operation: CFB-128 and OFB.
2. The performance thing mentioned in previous comments: it's not just about doing an `if`. The `if` does an additional register read from the peripheral over TL-UL. Unlike for other devices we have, AES operation is very quick: you get a new 128-bit block after 12 to 16 clock cycles. Saving even a single bus transaction can be significant.
3. ECB (sorry, we have been discussing a lot already about this): It turns out that for SCA, we primarily use ECB because SCA is mainly about leakage in the cipher core. So you don't want the other modes to obfuscate things. Re-Writing the key for every block is impractical and can become a performance bottleneck, especially as we're already in the millions of traces needed for analysis.

But as I said, we don't need to change this now. I think we shouldn't block this PR further and move on.

silvestrst commented:
Hey @vogelpi , thanks for the review. Also there is this #4833. The additional modes implementation will go to @alphan as a part of handover. I will see how many of the above comments it makes sense to fix-up myself and how many leave for Alphan as part of handover. I am planning on going through the comments today.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,582 @@
+   * Trigger encrypt/decrypt.
+   */
+  kDifAesTriggerStart = 0,
+  /**
+   * Clear key registers.
+   */
+  kDifAesTriggerKeyClear,
+  /**
+   * Clear Initialisation Vector (CBC), or Initial Counter Value (CTR)
+   * registers.
+   */
+  kDifAesTriggerIvClear,
+  /**
+   * Clear Input Data registers.
+   */
+  kDifAesTriggerDataInClear,
```
Thanks.

Done.

silvestrst commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,582 @@
+   * protection.
+   */
+  kDifAesStatusStall,
+  /**
```
One of us (me or Alphan will add it, when PR lands).

vogelpi replied:
This got merged now.

silvestrst replied:
Done.

mcy commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,589 @@
+ * @return `dif_aes_result_t`.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_aes_result_t dif_aes_status_get(const dif_aes_t *aes, dif_aes_status_t flag,
```
Hmmm, `get_status` for consistency with other DIFs?

silvestrst replied:
Done.

mcy commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,589 @@
+ * @param aes AES state data.
+ * @param flag Status flag to query.
+ * @param set Flag state (set/unset).
+ * @return `dif_aes_result_t`.
```
There seems to be a mix of `dif_aes_result_t` and "The result of the operation" values for this parameter. Maybe use one and be consistent? Same for the `aes` param.

silvestrst replied:
Done.

mcy commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -0,0 +1,424 @@
+typedef enum aes_mode_field_val {
+  kAesModeFieldValEcb = 0x01,  /**< The Electronic Codebook Mode. */
+  kAesModeFieldValCbc = 0x02,  /**< The Cipher Block Chaining Mode. */
+  kAesModeFieldValCfb = 0x04,  /**< TODO */
```
What does this TODO mean?

vogelpi replied:
It means these modes still need to be implemented. This is tracked in #4833.

mcy commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -0,0 +1,424 @@
+static void aes_clear_internal_state(const dif_aes_t *aes) {
+  // Make sure AES is idle before clearing.
+  while (!aes_idle(aes)) {
+  };
```
This semicolon is redundant.

silvestrst replied:
Done.

mcy commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -0,0 +1,424 @@
+    return kDifAesStartError;
+  }
+
+  // TODO - it is probably fine to clobber the CTRL register in process?
```
File a bug...?

silvestrst replied:
> File a bug...?

You mean an issue?

Anyhow - this has been removed.

mcy commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -0,0 +1,424 @@
+  }
+
+  // TODO - it is probably fine to clobber the CTRL register in process?
+  uint32_t reg =
+      mmio_region_read32(aes->params.base_addr, AES_CTRL_SHADOWED_REG_OFFSET);
+
+  reg =
+      bitfield_field32_write(reg, AES_CTRL_SHADOWED_KEY_LEN_FIELD, key_len_val);
+  reg = bitfield_field32_write(reg, AES_CTRL_SHADOWED_MODE_FIELD,
+                               cipher_mode_val);
```
Should this be a stanza?

silvestrst replied:
Not sure what you mean, I split up all `bitfield_field32_write` calls so that they are all on their own.

mcy commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -0,0 +1,424 @@
+                               cipher_mode_val);
+
+  // Default = bit unset (encrypt)
+  if (transaction->mode == kDifAesModeDecrypt) {
```
Why not inline this condition into the `true` below?

silvestrst replied:
I think it is fine either way, keeping as is.

mcy commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -0,0 +1,424 @@
+
+  // Default = bit unset (encrypt)
+  if (transaction->mode == kDifAesModeDecrypt) {
+    reg = bitfield_bit32_write(0, AES_CTRL_SHADOWED_OPERATION_BIT, true);
```
This `0` seems to imply that the previous manipulations to `reg` are discarded. Is this a bug?

silvestrst replied:
This is a bug, and If I saw this comment earlier, it would have saved me some time in debugging :)

mcy commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -0,0 +1,424 @@
+  return kDifAesOk;
+}
+
+dif_aes_result_t dif_aes_status_get(const dif_aes_t *aes, dif_aes_status_t flag,
+                                    bool *set) {
```
For a followup, it seems like this wants to return a bitset...?

silvestrst replied:
Possibly, I will leave it with @alphan to decide.

mcy commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -0,0 +1,424 @@
+    return kDifAesBadArg;
+  }
+
+  uint32_t index;
```
Nit: bitfield_bit_index_t.

silvestrst replied:
Done.

mcy commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,589 @@
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_warn_unused_result.h"
+
+// Header Extern Guard (so header can be used from C and C++)
```
Drop this comment.

silvestrst replied:
Done.

vogelpi commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -0,0 +1,426 @@
+        bitfield_bit32_write(reg, AES_CTRL_SHADOWED_MANUAL_OPERATION_BIT, true);
+  }
+
+  if (transaction->masking == kDifAesMaskingForceZero) {
+    reg =
+        bitfield_bit32_write(reg, AES_CTRL_SHADOWED_FORCE_ZERO_MASKS_BIT, true);
+  }
```
Thanks :+1:

vogelpi commented on sw/device/lib/dif/dif_aes.c:
```diff
@@ -0,0 +1,424 @@
+ * aes.CTRL.
+ */
+
+/*
+ * Field to select AES block cipher mode.
+ */
```
It seems you have accidentally removed the comment explaining the values and resolution of invalid values. Do you want to add that back?

silvestrst replied:
Done.

vogelpi commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,589 @@
+   * One share of the key that when XORed with `part1` results in the actual
+   * key.
+   */
+  uint32_t part0[8];
```
I think you wanted to change this to `share0` similarly `share1` below. Did you maybe forget to commit/push?

silvestrst replied:
Possibly, thanks.

Done.

vogelpi commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,589 @@
+   * Fatal alert conditions include i) storage errors in the Control Register,
+   * and ii) if any internal FSM enters an invalid state.
+   */
+  kDifAesAlertFatal = 0,
```
Really sorry to bother you again about this. I had to rename it again in #4909 . It's now `kDifAlertFatalFault` and `kDifAlertRecovCtrlUpdateErr`.

silvestrst replied:
Done.

vogelpi commented on sw/device/lib/dif/dif_aes.h:
```diff
@@ -0,0 +1,589 @@
+   * Fatal alert conditions include i) storage errors in the Control Register,
+   * and ii) if any internal FSM enters an invalid state.
+   */
+  kDifAesStatusAlertFatal,
```
Also these have changed again in line with the alert names. Sorry for that.

silvestrst replied:
Done.

vogelpi reviewed (commented):
Sorry for keeping you waiting @silvestrst . This looks mostly good but I think you forgot to commit/push some changes you've made (as you commented on them).

If the remaining things and also @mcy comments are addressed (I think all these are quick things) I am happy to approve this PR and move forward.

silvestrst reviewed (commented):
Thanks again guys for your review, I should have addressed all of the outstanding comments now.

I think we should try to merge this now unless there are major issues.

vogelpi reviewed (approved):
Thanks @silvestrst , this looks good to me!

To actually merge the PR, I think also @gkelly will need to approve it as he requested changes a long time ago. The interface still says: "Changes requested".
